src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs

```diff
@@ -3137,5 +3132,346 @@ static void M()
+
+            var position = source.GetRoot().DescendantNodes().OfType<MethodDeclarationSyntax>().Single().Position;
+
+            var x = speculative.GetRoot().DescendantNodes().OfType<AttributeSyntax>().Single();
+            var attributeConstructor = (SourceConstructorSymbol)semanticModel.GetSpeculativeSymbolInfo(position, x).Symbol;
+            Assert.Equal("A", attributeConstructor.ContainingType.Name);
+
+            Assert.True(semanticModel.TryGetSpeculativeSemanticModel(position, x, out var speculativeModel));
+            var attributeConstructor2 = speculativeModel.GetSymbolInfo(x).Symbol;
+            Assert.Equal("A", attributeConstructor2.ContainingType.Name);
+
+            comp.VerifyDiagnostics(
+                // (4,24): warning CS0169: The field 'Program.Unused' is never used
+                //     private static int Unused;
+                Diagnostic(ErrorCode.WRN_UnreferencedField, "Unused").WithArguments("Program.Unused").WithLocation(4, 24)
+                );
+        }
+
+        [Fact]
+        [WorkItem(16821, "https://github.com/dotnet/roslyn/issues/16821")]
+        public void AttributesMarkUsed()
+        {
+            var source = @"
+class A : System.Attribute
+{
+    public A(string x) { }
+}
+class Program
+{
+    static void Main()
+    {
+        int x;
+        void AttrParam([A(nameof(x))] int a) { }
+        void AttrParamLocfunc([A(nameof(AttrParam))] int a) { }
+        AttrParamLocfunc(2);
+        int y;
+        void AttrType<[A(nameof(y))] T>(int a) { }
+        void AttrTypeLocfunc<[A(nameof(AttrType))] T>(int a) { }
+        AttrTypeLocfunc<int>(2);
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib46(source, parseOptions: DefaultParseOptions, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics(
+                // (11,24): error CS8205: Attributes are not allowed on local function parameters or type parameters
+                //         void AttrParam([A(nameof(x))] int a) { }
+                Diagnostic(ErrorCode.ERR_AttributesInLocalFuncDecl, "[A(nameof(x))]").WithLocation(11, 24),
+                // (12,31): error CS8205: Attributes are not allowed on local function parameters or type parameters
+                //         void AttrParamLocfunc([A(nameof(AttrParam))] int a) { }
+                Diagnostic(ErrorCode.ERR_AttributesInLocalFuncDecl, "[A(nameof(AttrParam))]").WithLocation(12, 31),
+                // (15,23): error CS8205: Attributes are not allowed on local function parameters or type parameters
+                //         void AttrType<[A(nameof(y))] T>(int a) { }
+                Diagnostic(ErrorCode.ERR_AttributesInLocalFuncDecl, "[A(nameof(y))]").WithLocation(15, 23),
+                // (15,33): error CS0103: The name 'y' does not exist in the current context
+                //         void AttrType<[A(nameof(y))] T>(int a) { }
+                Diagnostic(ErrorCode.ERR_NameNotInContext, "y").WithArguments("y").WithLocation(15, 33),
```

Comment:
>Diagnostic(ErrorCode.ERR_NameNotInContext, "y").WithArguments("y").WithLocation(15, 33), [](start = 16, length = 88)

This doesn't feel right, it looks like we are not using the right binder to bind attributes on type parameters, please fix.